grpc/v1/webserver.proto

```diff
@@ -0,0 +1,625 @@
+
+  repeated FileMatch files = 3;
+
+  // RepoURLs holds a repo => template string map.
+  map<string, string> repo_urls = 4;
+
+  // FragmentNames holds a repo => template string map, for
+  // the line number fragment.
+  map<string, string> line_fragments = 5;
+}
+
+message SearchOptions {
+  // Return an upper-bound estimate of eligible documents in
+  // stats.ShardFilesConsidered.
+  bool estimate_doc_count = 1;
+
+  // Return the whole file.
+  bool whole = 2;
+
+  // Maximum number of matches: skip all processing an index
+  // shard after we found this many non-overlapping matches.
+  int64 shard_max_match_count = 3;
+
+  // Maximum number of matches: stop looking for more matches
+  // once we have this many matches across shards.
+  int64 total_max_match_count = 4;
+
+  // Maximum number of matches: skip processing documents for a repository in
+  // a shard once we have found ShardRepoMaxMatchCount.
+  //
+  // A compound shard may contain multiple repositories. This will most often
+  // be set to 1 to find all repositories containing a result.
+  int64 shard_repo_max_match_count = 5;
+
+  // Abort the search after this much time has passed.
+  google.protobuf.Duration max_wall_time = 6;
+
+  // FlushWallTime if non-zero will stop streaming behaviour at first and
+  // instead will collate and sort results. At FlushWallTime the results will
+  // be sent and then the behaviour will revert to the normal streaming.
+  google.protobuf.Duration flush_wall_time = 7;
+
+  // Trim the number of results after collating and sorting the
+  // results
+  int64 max_doc_display_count = 8;
+
+  // If set to a number greater than zero then up to this many number
+  // of context lines will be added before and after each matched line.
+  // Note that the included context lines might contain matches and
+  // it's up to the consumer of the result to remove those lines.
+  int64 num_context_lines = 9;
+
+  // If true, ChunkMatches will be returned in each FileMatch rather than LineMatches
+  // EXPERIMENTAL: the behavior of this flag may be changed in future versions.
+  bool chunk_matches = 10;
+
+  // EXPERIMENTAL. If true, document ranks are used as additional input for
+  // sorting matches.
+  bool use_document_ranks = 11;
+
+  // EXPERIMENTAL. When UseDocumentRanks is enabled, this can be optionally set to adjust
+  // their weight in the file match score. If the value is <= 0.0, the default weight value
+  // will be used. This option is temporary and is only exposed for testing/ tuning purposes.
+  double document_ranks_weight = 12;
+
+  // Trace turns on opentracing for this request if true and if the Jaeger address was provided as
+  // a command-line flag
+  bool trace = 13;
+
+  // If set, the search results will contain debug information for scoring.
+  bool debug_score = 14;
+
+  // SpanContext is the opentracing span context, if it exists, from the zoekt client
+  map<string, string> span_context = 15;
+}
+
+message Q {
```

Comment:
it feels like the query stuff should be in its own protofile, or is that inconvenient/unusual given it is used in the web server rpc?

Reply:
Certainly doable. Slightly inconvenient because then I need to figure out how to import 😄 Probably worth having an example in the codebase though

Reply:
Done in https://github.com/sourcegraph/zoekt/pull/577/commits/7e699c81da5cfdb9c9cf6e5dd53e12246b8830f9